### PR TITLE
Closes #367 `assert_*()` update cleanup

### LINKED ISSUE #367: Update messaging in `assert_*()` functions
When you're ready to tackle an one of the items below, convert it to an issue and self-assign. 

Generally, the following arguments will be added to each function. You can use `#' @inheritParams assert_logical_scalar` to add documentation for these new args.
- `arg_name = rlang::caller_arg(arg)`
- `message`
- `class`: This string will generally be the name of the assert function and will be passed to `cli::cli_abort(class = c(class, "assert-admiraldev")`
- `call = parent.frame()`

Use the following `{cli}` tags in the error messaging: `"{.arg xx}"` for argument names, `"{.val xx}"` for passed values, `"{.cls xx}"` for classes/types, and `{.obj_type_friendly xx}` for a friendly text description of an object type. Use other cli styling as needed. https://cli.r-lib.org/index.html

Unit tests must include at least one test for the correct error class, and one snapshot ensuring the exact messaging of the error condition. More details here https://docs.google.com/document/d/1EAKLQPis8XjcNsxjW0tmNj1A8H4Lw0UiryKAm1_6juo/edit?usp=sharing

DO THESE FIRST (these are used in the other `assert_*()` functions)
- [x] #368
- [x] #369
- [x] #372
- [x] #373
- [x] #375
- [x] #377

THEN DO THESE
- [x] #379
- [x] #380 
- [x] #385
- [x] #387
- [x] #389
- [x] #391
- [x] #396
- [x] #398
- [x] #403
- [x] #409
- [x] #400
- [x] #420
- [x] #395
- [x] #394
- [x] #414
- [x] #416 
- [x] #424
- [x] #422

FINALLY DO
- [x] Perform a final cleanup. Do we still need functions like `arg_name()`, `what_is_it()`, `compat_friendly_type.R`, `enumerate`? Anything else? Will removing these exported function result in downstream breaking changes?
- [x] Update `NEWS.md` that the messaging has been updated throughout, and note the breaking changes where the `assert_*(class)` argument has been renamed to `cls`

### LINKED ISSUE #437: Update `warn()` to use `cli::cli_warn()`
The last place the `enumerate()` is used is in one warn function. After that is updated, begin the deprecation process for it.

```r
  lifecycle::deprecate_warn(
    when = "1.1.0",
    what = "admiraldev::enumerate()",
    details = "This function was primarily used in error messaging, and can be
               replaced with 'cli' functionality `cil::cli_abort('{.val {letters[1:3]}}')`"
  )
```

### PR DESCRIPTION
**Pull request by ddsjoberg**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/dev/CONTRIBUTING.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #367 and Closes #437 into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiraldev/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -4,12 +4,22 @@
 
 ## Updates of Existing Functions
 
+* Error messaging throughout the package have been updated from `rlang::abort()` to `cli::cli_abort()`. As a part of the update, each of the `assert_*()` functions have new arguments `assert_*(message, arg_name, call, class).` (#367)
+
+* Warning messaging has also been updated to use {cli} messaging.
```
Comment (bms63) [inline NEWS.md:9]:
```suggestion
* Warning messaging has also been updated to use ‘{cli}’ messaging.
```

Comment (bms63) [reply NEWS.md:9]: That is supposed to backticks , but I’m on my phone

File: R/what.R [hunk 2]
```diff
@@ -18,6 +18,13 @@
 #' what_is_it(1:10)
 #' what_is_it(mtcars)
 what_is_it <- function(x) {
```
Comment (bms63) [inline R/what.R:20]: deprecate_stop ??

File: NEWS.md (hunk 1 above)
Comment (ddsjoberg) [reply NEWS.md]: updated

File: R/what.R (hunk 2 above)
Comment (ddsjoberg) [reply R/what.R:20]: The more I thought about it, the more I felt that there was nothing to lose by including one cycle of a warn deprecation 🤷🏼  We can hard deprecate after the release?

Comment (bms63) [reply R/what.R:20]: Okay sounds good!   I thought I was being the bad guy and deviating from our process, but here is Daniel following admiral byzantine processes.  Have we won?

File: tests/testthat/test-warnings.R [hunk 3]
```diff
@@ -3,17 +3,14 @@
 test_that("warn_if_vars_exist Test 1: warning if a variable already exists in the input dataset", {
   dm <- pharmaversesdtm::dm
```
Comment (bms63) [inline tests/testthat/test-warnings.R:4]: oohhh can we make a dummy example remove pharmaversesdtm from the Suggest?

File: R/what.R (hunk 2 above)
Comment (ddsjoberg) [reply R/what.R:20]: HA! It's possible!

File: tests/testthat/test-warnings.R (hunk 3 above)
Comment (ddsjoberg) [reply tests/testthat/test-warnings.R:4]: I think @millerg23 is tackling that in #434

**Review by bms63 — APPROVED** [review]
Thanks again @ddsjoberg !!
